nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/AppLibClassLoaderServiceImpl.java

```diff
@@ -129,6 +120,22 @@ public DelegatingClassLoader run() {
         return (DelegatingClassLoader.ClassFinder)appLibClassFinder;
     }
 
+    private static class AppLibClassLoader extends DelegatingClassLoader implements Closeable {
+
+        public AppLibClassLoader(ClassLoader parent) {
+            super(parent);
+        }
+
+        @Override
+        public void close() throws IOException {
+            for (ClassFinder classFinder : getDelegates()) {
+                if (classFinder instanceof URLClassFinder) {
+                    ((GlassfishUrlClassLoader) classFinder).close();
```

Comment:
Could the types `instanceof`ed and cast be synchronized?

Reply:
I don't think it's necessary. Instead, delegates list should be properly synchronized.

Reply:
by _synchronized_ I meant: to use one type, like `GlassfishUrlClassLoader` (not like MT-`synchronized`)

Reply:
I apologize for the misunderstanding. No, we cannot. Delegates list may contain 2 types of `ClassFinder's` and both of them are `Closeable` (one extends `GlassFishClassLoader` directry, second - indrectly (extends `ASURLClassLoader`)). The only thing we can do - cast to `URLClassLoader` or even to `Closeable` instead of `GlassFishUrlClassLoader`.

Reply:
Using `T := URLClassFinder`:
```java
                 if (classFinder instanceof URLClassFinder) {
                    ((URLClassFinder) classFinder).close();
                 }
```
I can't see what's wrong with that.

Reply:
Hmm... Yo're right)) Thanks.